### PR TITLE
[SYCL] Runtime property lists on kernel invocation functions

### PR DESCRIPTION
**Pull request by againull**
Add an optional parameter to kernel invocation functions which accepts
property list. Such properties are going to affect the way the kernel
is invoked. Example is a property to set cache/SLM size for a kernel
invocation which is going to be implemented in upcoming PR.

Test: https://github.com/intel/llvm-test-suite/pull/197
Draft PR for GPU Cache Config property: https://github.com/intel/llvm/pull/3417

### CODE REVIEW

**Comment by illuhad** [comment]
I would suggest to be careful with adding more overloads to the kernel invocation functions for this functionality. They are already exceedingly complicated because of reductions, and I expect they will become even more complicated in the future with the addition of a library-only capable replacement mechanism for kernel attributes.
Because of this, we have instead decided to attach a `property_list` to `queue::submit()` already some time ago in hipSYCL to provide similar functionality:
https://github.com/illuhad/hipSYCL/blob/develop/doc/extensions.md#hipsycl_ext_cg_property_-command-group-properties

This also retains the current convention that the kernel always comes as last argument of the invocation function, which I personally prefer for aesthetic reasons ;)

File: sycl/test/abi/layout_handler.cpp [hunk 1]
```diff
@@ -186,5 +186,19 @@ void foo() {
 // CHECK-NEXT: 544 |         std::__shared_ptr<class sycl::detail::event_impl, __gnu_cxx::_S_atomic>::element_type * _M_ptr
 // CHECK-NEXT: 552 |         class std::__shared_count<__gnu_cxx::_S_atomic> _M_refcount
 // CHECK-NEXT: 552 |           _Sp_counted_base<(enum __gnu_cxx::_Lock_policy)2U> * _M_pi
-// CHECK-NEXT:     | [sizeof=560, dsize=560, align=8,
-// CHECK-NEXT:     |  nvsize=560, nvalign=8]
```
Comment (s-kanaev) [inline sycl/test/abi/layout_handler.cpp:190]: Isn't breaking ABI avoidable?
Tagging @romanovvlad .

Comment (romanovvlad) [reply sycl/test/abi/layout_handler.cpp:190]: Yeah, Artur, please avoid breaking ABI. Similar problem is being solved here: https://github.com/intel/llvm/pull/3375

**Comment by againull** [comment]
This PR is going to be redone.

**Pull request closed without merging**